File: sentinelai-frontend/src/types/soar.ts
export interface Playbook {
  id: string;
  name: string;
  description: string | null;
  playbook_type: string;
  severity: string;
  category: string | null;
  tags: string[] | null;
  is_template: boolean;
  is_active: boolean;
  steps: PlaybookStep[];
  config: Record<string, any> | null;
  execution_count: number;
  avg_execution_time: number;
  success_rate: number;
  created_at: string;
  updated_at: string;
  created_by: string | null;
}

export interface PlaybookListItem {
  id: string;
  name: string;
  playbook_type: string;
  severity: string;
  category: string | null;
  is_template: boolean;
  is_active: boolean;
  execution_count: number;
  success_rate: number;
  created_at: string;
}

export interface PlaybookStep {
  index: number;
  name: string;
  action: string;
  config: Record<string, any>;
  type: 'action' | 'condition' | 'delay' | 'approval' | 'loop';
}

export interface PlaybookExecution {
  id: string;
  playbook_id: string;
  playbook_name: string;
  incident_id: string | null;
  status: string;
  current_step: number;
  total_steps: number;
  started_at: string | null;
  completed_at: string | null;
  duration_ms: number | null;
  triggered_by: string | null;
  error_message: string | null;
  failed_step: number | null;
  execution_data: Record<string, any> | null;
  ai_summary: string | null;
  created_at: string;
}

export interface PlaybookExecutionLog {
  id: string;
  execution_id: string;
  step_index: number;
  step_name: string;
  action_type: string;
  status: string;
  message: string | null;
  duration_ms: number | null;
  input_data: Record<string, any> | null;
  output_data: Record<string, any> | null;
  error: string | null;
  created_at: string;
}

export interface ApprovalRequest {
  id: string;
  playbook_id: string;
  execution_id: string;
  incident_id: string | null;
  step_index: number;
  action_type: string;
  requested_by: string | null;
  approved_by: string | null;
  status: string;
  reason: string | null;
  resolved_at: string | null;
  created_at: string;
}

export interface PlaybookStats {
  total_playbooks: number;
  active_playbooks: number;
  executions_today: number;
  total_executions: number;
  success_rate: number;
  avg_execution_time_ms: number;
  pending_approvals: number;
  by_type: Record<string, number>;
  by_severity: Record<string, number>;
  recent_executions: { id: string; playbook_name: string; status: string; created_at: string }[];
}

export interface ActionDefinition {
  name: string;
  description: string;
}

export interface TemplateDefinition {
  type: string;
  steps: number;
  name: string;
}

export interface ExecuteResponse {
  execution_id: string;
  status: string;
  message: string;
}

export interface PaginatedData<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  has_next: boolean;
  has_prev: boolean;
}

export interface ApiResponse<T> {
  success: boolean;
  message: string;
  data: T;
}

File: sentinelai-backend/app/schemas/soar.py
from datetime import datetime
from typing import Any

from pydantic import Field
from app.schemas.base import BaseSchema


class PlaybookCreate(BaseSchema):
    name: str
    description: str | None = None
    playbook_type: str
    severity: str = "medium"
    category: str | None = None
    tags: list[str] | None = None
    is_template: bool = False
    steps: list[dict[str, Any]] = []
    config: dict[str, Any] | None = None


class PlaybookUpdate(BaseSchema):
    name: str | None = None
    description: str | None = None
    playbook_type: str | None = None
    severity: str | None = None
    category: str | None = None
    tags: list[str] | None = None
    is_active: bool | None = None
    is_template: bool | None = None
    steps: list[dict[str, Any]] | None = None
    config: dict[str, Any] | None = None


class PlaybookResponse(BaseSchema):
    id: str
    name: str
    description: str | None = None
    playbook_type: str
    severity: str
    category: str | None = None
    tags: list[str] | None = None
    is_template: bool = False
    is_active: bool = True
    steps: list[dict[str, Any]] | None = None
    config: dict[str, Any] | None = None
    execution_count: int = 0
    avg_execution_time: float = 0.0
    success_rate: float = 0.0
    created_at: datetime
    updated_at: datetime
    created_by: str | None = None


class PlaybookListItem(BaseSchema):
    id: str
    name: str
    playbook_type: str
    severity: str
    category: str | None = None
    is_template: bool = False
    is_active: bool = True
    execution_count: int = 0
    success_rate: float = 0.0
    created_at: datetime


class PlaybookExecutionCreate(BaseSchema):
    playbook_id: str
    incident_id: str | None = None
    triggered_by: str | None = None


class PlaybookExecutionResponse(BaseSchema):
    id: str
    playbook_id: str
    playbook_name: str
    incident_id: str | None = None
    status: str
    current_step: int = 0
    total_steps: int = 0
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = None
    triggered_by: str | None = None
    error_message: str | None = None
    failed_step: int | None = None
    execution_data: dict[str, Any] | None = None
    ai_summary: str | None = None
    created_at: datetime


class PlaybookExecutionLogResponse(BaseSchema):
    id: str
    execution_id: str
    step_index: int
    step_name: str
    action_type: str
    status: str
    message: str | None = None
    duration_ms: int | None = None
    input_data: dict[str, Any] | None = None
    output_data: dict[str, Any] | None = None
    error: str | None = None
    created_at: datetime


class ApprovalRequestCreate(BaseSchema):
    execution_id: str
    step_index: int
    action_type: str
    reason: str | None = None


class ApprovalRequestResponse(BaseSchema):
    id: str
    playbook_id: str
    execution_id: str
    incident_id: str | None = None
    step_index: int
    action_type: str
    requested_by: str | None = None
    approved_by: str | None = None
    status: str
    reason: str | None = None
    resolved_at: datetime | None = None
    created_at: datetime


class ApprovalAction(BaseSchema):
    approve: bool = True
    reason: str | None = None


class PlaybookStats(BaseSchema):
    total_playbooks: int = 0
    active_playbooks: int = 0
    executions_today: int = 0
    total_executions: int = 0
    success_rate: float = 0.0
    avg_execution_time_ms: float = 0.0
    pending_approvals: int = 0
    by_type: dict[str, int] = {}
    by_severity: dict[str, int] = {}
    recent_executions: list[dict[str, Any]] = []


class ExecuteResponse(BaseSchema):
    execution_id: str
    status: str
    message: str
